flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -347,18 +352,39 @@ private void handleStartTaskExecutorServicesException(Exception e) throws Except
 
 		resourceManagerHeartbeatManager.stop();
 
-		try {
-			stopTaskExecutorServices();
-		} catch (Exception e) {
-			throwable = ExceptionUtils.firstOrSuppressed(e, throwable);
-		}
+		List<CompletableFuture<Void>> taskCompletionFutures = failSubmittedTasks();
 
 		if (throwable != null) {
+			try {
+				stopTaskExecutorServices();
+			} catch (Exception e) {
+				throwable = ExceptionUtils.firstOrSuppressed(e, throwable);
+			}
 			return FutureUtils.completedExceptionally(new FlinkException("Error while shutting the TaskExecutor down.", throwable));
 		} else {
-			log.info("Stopped TaskExecutor {}.", getAddress());
-			return CompletableFuture.completedFuture(null);
+			return FutureUtils.waitForAll(taskCompletionFutures).thenRun(() -> {
+				try {
+					stopTaskExecutorServices();
+				} catch (Exception e) {
+					throw new FlinkRuntimeException("Error while shutting the TaskExecutor down.", e);
+				}
+				log.info("Stopped TaskExecutor {}.", getAddress());
+			});
```

Comment:
I think @tillrohrmann (correct me if not) meant `If throwable == null` then still try `stopTaskExecutorServices`.
Then we should register this callback for the result future, something like:
```
...
})
.handle((v, t) -> {
  ...
      stopTaskExecutorServices();
    } catch (Exception e) {
	  stopFailure = new FlinkRuntimeException("Error while shutting the TaskExecutor down.", e);
      throwable = ExceptionUtils.firstOrSuppressed(stopFailure, throwable);
	}
  ...
  if (t != null) { 
    t = ExceptionUtils.firstOrSuppressed(t, throwable);
    throw t; 
  } else if (throwable !=null) {
    throw throwable;
  }
});
```
instead of `if (throwable != null) {...} else ...` and `thenRun`

Reply:
I'm afraid that the `throwable` can't be used in the callback because it is not a `final` var.

Reply:
It can be copied to a final var:
```
final Throwable throwableBeforeTasksCompletion = jobManagerDisconnectThrowable;
return taskCompletionTracker.failIncompleteTasksAndWaitForAllCompleted()
	.handle((v, tasksCompletionThrowable) -> {
		Throwable resultThrowable = tasksCompletionThrowable == null ? null :
			ExceptionUtils.firstOrSuppressed(tasksCompletionThrowable, throwableBeforeTasksCompletion);
		try {
			stopTaskExecutorServices();
		} catch (Exception e) {
			resultThrowable = ExceptionUtils.firstOrSuppressed(e, resultThrowable);
		}
		if (resultThrowable != null) {
			throw new FlinkRuntimeException("Error while shutting the TaskExecutor down.", resultThrowable);
		}
		log.info("Stopped TaskExecutor {}.", getAddress());
		return null;
	});
```